### PR TITLE
Disable topk optimisation on dispatch when content distribution is se…

### PR DESCRIPTION
**Pull request by baldersheim**
…verly skewed.

When the skew is too large the assumption that docs are evenly and randomly distributed hold.
The impact and is larger on smaller systems. In large systems the where this optimisation is more important,
the probabilitity of large skew will be less.

@jonmv PR

### CODE REVIEW

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java [hunk 1]
```diff
@@ -53,25 +55,30 @@ void setHasSufficientCoverage(boolean sufficientCoverage) {
...
+            long deviation = nodes.stream().filter(node -> node.isWorking() == Boolean.TRUE).mapToLong(node -> Math.abs(node.getActiveDocuments() - average)).sum();
+            isContentWellBalanced.set(deviation <= (activeDocs * MAX_UNBALANCE));
+        } else {
+            isContentWellBalanced.set(true);
+        }
+
     }
 
-    /** Returns the active documents on this node. If unknown, 0 is returned. */
+    /** Returns the active documents on this groip. If unknown, 0 is returned. */
```
Comment (jonmv) [inline container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java:76]:
```suggestion
    /** Returns the active documents on this group. If unknown, 0 is returned. */
```

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java [hunk 2]
```diff
@@ -53,25 +55,30 @@ void setHasSufficientCoverage(boolean sufficientCoverage) {
...
 
     public int workingNodes() {
-        int nodesUp = 0;
-        for (Node node : nodes) {
-            if (node.isWorking() == Boolean.TRUE) {
-                nodesUp++;
-            }
-        }
-        return nodesUp;
+        return (int) nodes.stream().filter(node -> node.isWorking() == Boolean.TRUE).count();
```
Comment (jonmv) [inline container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java:58]:
```suggestion
        return (int) nodes.stream().filter(node -> node.isWorking()).count();
```

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java [hunk 3]
```diff
@@ -53,25 +55,30 @@ void setHasSufficientCoverage(boolean sufficientCoverage) {
...
     }
 
     void aggregateNodeValues() {
-        activeDocuments.set(nodes.stream().filter(node -> node.isWorking() == Boolean.TRUE).mapToLong(Node::getActiveDocuments).sum());
-        isBlockingWrites.set(nodes.stream().anyMatch(node -> node.isBlockingWrites()));
+        long activeDocs = nodes.stream().filter(node -> node.isWorking() == Boolean.TRUE).mapToLong(Node::getActiveDocuments).sum();
```
Comment (jonmv) [inline container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java:62]:
```suggestion
        long activeDocs = nodes.stream().filter(node -> node.isWorking()).mapToLong(Node::getActiveDocuments).sum();
```

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java [hunk 4]
```diff
@@ -53,25 +55,30 @@ void setHasSufficientCoverage(boolean sufficientCoverage) {
...
+        activeDocuments.set(activeDocs);
+        isBlockingWrites.set(nodes.stream().anyMatch(Node::isBlockingWrites));
+        int numWorkingNodes = workingNodes();
+        if (numWorkingNodes > 0) {
+            long average = activeDocs / numWorkingNodes;
+            long deviation = nodes.stream().filter(node -> node.isWorking() == Boolean.TRUE).mapToLong(node -> Math.abs(node.getActiveDocuments() - average)).sum();
```
Comment (jonmv) [inline container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java:68]:
```suggestion
            long deviation = nodes.stream().filter(node -> node.isWorking()).mapToLong(node -> Math.abs(node.getActiveDocuments() - average)).sum();
```

**Review by jonmv — APPROVED** [review]
Just some details. Nice! What about nodes with duplicate corpuses?

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java (hunk 2 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java:58]: What happens if 'null' is returned ? Wont that give NullPointerException ?

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java (hunk 3 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java:62]: See comment above

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java (hunk 4 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java:68]: See comment above

**Comment by baldersheim** [comment]
Node.getActiveDocuments() only counts the primary copy which is the only one that is considered.

**Comment by jonmv** [comment]
Ouch, `Boolean` return value! You're right, that would be an NPE. 

How do the primary copies change during redistribution from a retiring node?

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java (hunk 2 above)
Comment (jonmv) [reply container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Group.java:58]: 💥

**Comment by baldersheim** [comment]
Once a bucket has been copied to a new node, the copy which is now primary will be activated and the previous will be deactivated. Only active copies are searched. For a short period of time(milliseconds) there might be duplicates, but that is filtered higher up in the stack based on the 96bit GID.

**Comment by jonmv** [comment]
So if all content nodes in a group are concurrently migrating from a set of old nodes to a set of new ones, and they're all halfway ±5%, then this will enable the optimisation anyway? Only in this scenario? If it _is_ enabled in this scenario, we'll get wrong results, assuming the search considers not just the primaries on a node (which I believe it does?).

**Comment by baldersheim** [comment]
Yes once they are halfway it will be enabled again, and results will be correct.
Only 1 copy of a bucket is searched, except for the small subset of buckets being activated/deactivated.

**Comment by jonmv** [comment]
Ok, perfect :)

**Pull request merged**